File: 2.go
/**
demo for example 1 and 2
*/

package main

import (
	"fmt"
	"time"
)

var n int = 64 * 1024 * 1024

func main() {
	for i := 0; i <= 15; i++ {
		t(1 << i)
	}
}

func t(step int) {
	//var r int
	arr := make([]int, n)
	for i, _ := range arr {
		arr[i] = i
	}
	fmt.Printf("test %d\t", step)
	s := time.Now().UnixNano()
	for i := 0; i < n; i += step {
		//r ^= arr[i]
		arr[i] *= 3
	}
	//println(r)
	e := time.Now().UnixNano()
	fmt.Println(float32(e-s) / 1000000)
}

File: 2.c
#include <assert.h>
#include <sys/time.h>
#include <stdio.h>
#include <stdlib.h>

const int n =  64 * 1024 * 1024;

void t(int step){
    int *arr = (int*)malloc(sizeof(int) * n);
    assert(arr != NULL);

    printf("test %d\t", step);

    struct timeval tv;
    gettimeofday(&tv,NULL);
    int s = tv.tv_sec;
    int su = tv.tv_usec;

    for (int i = 0; i < n; i+=step) arr[i] *= 3;

    gettimeofday(&tv,NULL);
    int e = tv.tv_sec;
    int eu = tv.tv_usec;
    printf("%.4f\n", (e-s)+(eu-su)/1000000.0);

    free(arr);
}

int main(){
    for (int i = 0; i <= 15; i+=1) t(1 << i);
}
